File: lib/validateNames.ts
/**
 * Validates Names of Allah returned by Claude against the canonical 99 Names list.
 * Prevents hallucinated or non-standard names from appearing in the app.
 */
import { ALLAH_NAMES } from '@/constants/allahNames';

// Build a normalised lookup set from the canonical list
// Normalise: lowercase, remove diacritics, remove "al-"/"ar-"/"as-"/"ash-" prefix variation
function normalise(name: string): string {
  return name
    .toLowerCase()
    .replace(/[\u0600-\u06FF\s]/g, '') // strip Arabic script
    .replace(/[^a-z]/g, '')            // strip non-alpha
    .replace(/^(al|ar|as|ash|at|az|an)/, ''); // strip common prefixes
}

const CANONICAL_NORMALISED = new Set(
  ALLAH_NAMES.map(n => normalise(n.transliteration))
);

// Also build a map from normalised -> canonical entry for lookup
const CANONICAL_MAP = new Map(
  ALLAH_NAMES.map(n => [normalise(n.transliteration), n])
);

export function isValidAllahName(name: string): boolean {
  return CANONICAL_NORMALISED.has(normalise(name));
}

/**
 * Given a name string returned by Claude, find the closest canonical match.
 * Returns the canonical entry if found, null if it's a hallucination.
 */
export function findCanonicalName(name: string): { name: string; nameArabic: string } | null {
  const key = normalise(name);
  const entry = CANONICAL_MAP.get(key);
  if (entry) return { name: entry.transliteration, nameArabic: entry.arabic };
  return null;
}

/**
 * Filter a list of {name, nameArabic, ...} objects to only include valid Names.
 * Replaces name/nameArabic with canonical values where a match is found.
 */
export function filterValidNames<T extends { name: string; nameArabic: string }>(
  names: T[]
): T[] {
  return names.reduce<T[]>((acc, item) => {
    const canonical = findCanonicalName(item.name);
    if (canonical) {
      acc.push({ ...item, name: canonical.name, nameArabic: canonical.nameArabic });
    }
    return acc;
  }, []);
}

/**
 * Build a compact canonical names reference string to inject into Claude prompts.
 * This constrains Claude to only use real Names.
 */
export function buildCanonicalNamesPromptList(): string {
  return ALLAH_NAMES.map(n => `${n.transliteration} (${n.arabic}) — ${n.english}`).join('\n');
}

File: flutter/lib/services/validate_names.dart
/// Validates Names of Allah returned by Claude against the canonical 99 Names list.
/// Prevents hallucinated or non-standard names from appearing in the app.
library;

import 'package:sakina/core/constants/allah_names.dart';

/// Normalise: lowercase, remove Arabic script, remove non-alpha,
/// strip common transliteration prefixes (al-, ar-, as-, ash-, at-, az-, an-).
String _normalise(String name) {
  return name
      .toLowerCase()
      .replaceAll(RegExp(r'[\u0600-\u06FF\s]'), '') // strip Arabic script
      .replaceAll(RegExp(r'[^a-z]'), '') // strip non-alpha
      .replaceFirst(RegExp(r'^(al|ar|as|ash|at|az|an)'), ''); // strip common prefixes
}

/// Pre-built normalised lookup set from the canonical list (including aliases).
final Set<String> _canonicalNormalised = _canonicalMap.keys.toSet();

/// Common AI transliteration variants → canonical Name. The AI sometimes
/// returns vowel-length variants (Al-Wakil vs canonical Al-Wakeel) or
/// dh↔z variants (Al-Dhahir vs canonical Az-Zahir). These map to the same
/// underlying Arabic Name, but `_normalise` doesn't collapse them because
/// a blanket `ee→i` rule would conflate Al-Majeed and Al-Majid (two
/// distinct Names). Keep this map small and only add entries the eval surfaces.
const Map<String, String> _transliterationAliases = {
  'Al-Wakil': 'Al-Wakeel',
  'Al-Dhahir': 'Az-Zahir',
  'Al-Halim': 'Al-Haleem',
  'Al-Latif': 'Al-Lateef',
};

/// Pre-built map from normalised key to canonical entry. Includes aliases
/// so AI-returned variants resolve to the same `AllahName` as their canonical form.
final Map<String, AllahName> _canonicalMap = () {
  final m = <String, AllahName>{
    for (final n in allahNames) _normalise(n.transliteration): n,
  };
  for (final entry in _transliterationAliases.entries) {
    final canonical = allahNames.firstWhere(
      (n) => n.transliteration == entry.value,
      orElse: () => throw StateError(
          'alias target ${entry.value} not in allahNames'),
    );
    m[_normalise(entry.key)] = canonical;
  }
  return m;
}();

/// Returns true if [name] matches a canonical Name of Allah.
bool isValidAllahName(String name) {
  return _canonicalNormalised.contains(_normalise(name));
}

/// Given a name string returned by Claude, find the closest canonical match.
/// Returns a record with name and nameArabic if found, null if it's a hallucination.
({String name, String nameArabic})? findCanonicalName(String name) {
  final key = _normalise(name);
  final entry = _canonicalMap[key];
  if (entry != null) {
    return (name: entry.transliteration, nameArabic: entry.arabic);
  }
  return null;
}

/// Filter a list of maps to only include valid Names.
/// Replaces name/nameArabic with canonical values where a match is found.
/// Each item must have 'name' and 'nameArabic' keys.
List<Map<String, dynamic>> filterValidNames(List<Map<String, dynamic>> names) {
  return names.fold<List<Map<String, dynamic>>>([], (acc, item) {
    final canonical = findCanonicalName(item['name'] as String);
    if (canonical != null) {
      acc.add({
        ...item,
        'name': canonical.name,
        'nameArabic': canonical.nameArabic,
      });
    }
    return acc;
  });
}

/// Pre-built canonical names reference string. Compile-time once (the underlying
/// `allahNames` is const), reused across every AI call. Previously was a function
/// that rebuilt the same ~3KB string on every invocation (4 sites per reflect/
/// daily/findNames call) — pure waste since the input is const.
final String _canonicalNamesPromptList = allahNames
    .map((n) => '${n.transliteration} (${n.arabic}) — ${n.english}')
    .join('\n');

/// Build a compact canonical names reference string to inject into Claude prompts.
/// This constrains Claude to only use real Names.
String buildCanonicalNamesPromptList() => _canonicalNamesPromptList;
